DOCS-10596-jms-topic-subs-kt (#1409)
mulesoft/docs-connectors: jms/1.7/modules/ROOT/pages/jms-topic-subscription.adoc

= Configure Topic Subscriptions with JMS Connector
:keywords: jms, connector, consume, source, listener, topic, subscription
:page-aliases: connectors::jms/jms-topic-subscription.adoc

> Topics in JMS enables you to configure how the subscription behaves regarding the connection of the consumer and also how these consumers interact with each other.

Reviewer: Because "topics" is plural, the verb must agree: topics enable, not topics enables.

Reviewer: So this confused me: a topic enables you to configure something?
Should we say "the use of topics enables you"?
Or do we mean that topic subscriptions enable you to, or "the use of topic subscriptions enables you"?

Reviewer: The connection of the consumer to what?
Also, we should say "the connection of the consumer to <> and how multiple consumers interact."

Reviewer: copyedit:
```suggestion
Topic subscriptions in JMS enable you to configure how the subscription behaves regarding the connection of  consumers and how consumers interact.
```

> The Anypoint Connector for JMS (JMS Connector) *Consumer* and *On New Message* operations can manage the aspects of a topic subscription in the *Consumer type* field by selecting the *Topic consumer* configuration.

Reviewer: Is this accurate?
```suggestion
The Anypoint Connector for JMS (JMS Connector) *Consumer* and *On New Message* operations manage topic subscription based on whether you set the *Consumer type* field of *Topic consumer* to *Durable*, *Shared*, or *No local*.
```

== Configure Durable Subscriptions

> Durable subscriptions enables you to receive published messages to the topic while the consumer is offline.

Reviewer: subject + verb agreement:
```suggestion
Durable subscriptions enable you to receive messages published to the topic while the consumer is offline.
```

> Once you configure a durable subscription, an available consumer receives all the published messages before the established connection. +

Reviewer: GLOBALLY replace "once" with either "after" or "one time," depending on your intent (translation):
```suggestion
After you configure a durable subscription, an available consumer receives all the published messages before the established connection. +
```
Isn't this just repeating the previous sentence? Are we trying to clarify that durable subscriptions enable all subscribed consumers to receive one another's published messages, even without a current connection?

> Then the JMS provider stores the messages sent or published to the topic, as it would store messages sent to a queue. If another application has a topic consumer configuration that uses the same connection and its client ID, the same topic, and the same subscription name, the subscription reactivates, and the JMS provider delivers the published messages while the subscriber was inactive.

Reviewer: To what does "Then" refer? What comes before this to make this later?

Reviewer: Best guess:
```suggestion
The JMS provider stores messages sent or published to a topic, just as it would messages sent to a queue. If another consumer uses the same connection, client ID, topic, and subscription name, the JMS provider delivers the published messages, even if while the subscriber is inactive.
```
But I sort of want an illustration here.
Are we using "consumer", "topic consumer configuration," and "subscriber" to mean the same thing here: the recipient of the message? Are the messages "sent" or "published"? Supplying alternative terms makes it difficult, at least for me, to follow the narrative.

In the following example, you configure the *Durable* field for the *Topic consumer* configuration:

. In Studio, select the *Consume* operation from your flow.
. Set the *Destination* field to `#[vars.destination]`.

> . Set the *Consumer type* field to *Topic consumer*:

Reviewer: ```suggestion
. Set the *Consumer type* field to *Topic consumer*.
```

</jms:consume>
----

== Configure Shared Subscriptions

In a JMS spec 1.1 connection, a subscription to a topic cannot have more than one consumer at a time, meaning that the processing work of messages on a topic subscription cannot be shared, thereby limiting the scalability of the application.

> JMS spec 2.0 connection removes this restriction and enables you to configure shared consumers on the same subscription.

Reviewer 2: I think this should be "The JMS spec 2.0 connection"

</jms:consume>
----

== Configure Shared Subscriptions

> In JMS spec 1.1 connection, a subscription on a topic can not have more than one consumer at a time, meaning that processing work of messages on a topic subscription can not be shared, thereby limiting the scalability of the application.

Reviewer: copyedit:
```suggestion
In a JMS spec 1.1 connection, a subscription to a topic cannot have more than one consumer at a time, meaning that the processing work of messages on a topic subscription cannot be shared, thereby limiting the scalability of the application.
```
Best guess:
```suggestion
In a JMS spec 1.1 connection, only one consumer at a time can access a subscription to a topic, disallowing shared message processing and limiting the scalability of the application.
```

> JMS spec 2.0 connection removes this restriction and enables you to configure shared consumers on the same subscription.

Reviewer: ```suggestion
JMS spec 2.0 connection removes this restriction and enables you to configure shared subscription consumption.
```

> In the following example, you configure the *Shared* field for the *Topic consumer* configuration. With this configuration on two different *On New Messages* source instances, when a message is published to the invoices destination it is delivered only to one of the sources and not the other:

Reviewer: This wording is used earlier and I am not certain that it is precise. I think we are overusing "configure" and "configuration," to the extent that it impairs meaning. Should this really be...:
```suggestion
In the following example, you set *Topic consumer* to *Shared*.
```
The rest of this intro is confusing, because it refers to "two different *On New Messages* " instances that are not shown next, and it describes a NOT sharing situation.

</jms:listener>
----

== Configure No Local Subscriptions

> There are cases in which you may not want to receive the published messages to a topic by the same application. To prevent the consumer to receive published messages to the topic by the same connection, configure the *Topic consumer* field to *No local*. Consumers configured as no local can not be part of a shared subscription.

Reviewer: Yeah...I think you need to have a peer look at this whole topic, because I am lost.
My guess for this text is...:
```suggestion
If you want to limit the published messages available to a consumer to only those from outside connections, set *Topic consumer* to *No local*.  Consumers configured as such cannot be part of a shared subscription.
```

> In the following example, you configure the *No local* field for the *Topic consumer* configuration:

Reviewer: ```suggestion
In the following example, you set *Topic consumer* to "No local":
```

. In Studio, select the *Consume* operation from your flow.
. Set the *Destination* field to `{notificationsChannel}`.
. Set the *Consumer type* field to *Topic consumer*:

> . Select the *Shared* option.

Reviewer: This is a typo, right?
```suggestion
. Select the *No local* option.
```

</jms:listener>
----

== See Also

> * xref:jms-listener.adoc[Listen For New Messages]

Reviewer: ```suggestion
* xref:jms-listener.adoc[Listen for New Messages]
```
